File: sdk/python/polar/checkouts.py
from __future__ import annotations

import builtins
import typing

from polar.base import AsyncServiceBase, SyncServiceBase, parse_response
from polar.errors import (
    ClientConfirm403Error,
    ClientUpdate403Error,
    ExpiredCheckoutError,
    HTTPValidationError,
    PaymentError,
    ResourceNotFound,
    Update403Error,
)
from polar.inputs import (
    CheckoutConfirmStripe,
    CheckoutCreate,
    CheckoutUpdate,
    CheckoutUpdatePublic,
)
from polar.literals import (
    CheckoutSortProperty,
    CheckoutStatus,
)
from polar.outputs import (
    Checkout,
    CheckoutPublic,
    CheckoutPublicConfirmed,
    ListResourceCheckout,
)


class CheckoutsSync(SyncServiceBase):
    def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        status: CheckoutStatus | builtins.list[CheckoutStatus] | None = None,
        query: str | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutSortProperty] | None = ["-created_at"],
    ) -> ListResourceCheckout:
        """
        List checkout sessions.

        **Scopes**: `checkouts:read` `checkouts:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by customer external ID.
            status: Filter by checkout session status.
            query: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "product_id": product_id,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "status": status,
                "query": query,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceCheckout, method_errors)

    def create(
        self,
        **kwargs: typing.Unpack[CheckoutCreate],
    ) -> Checkout:
        """
        Create a checkout session.

        **Scopes**: `checkouts:write`

        Args:

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkouts/",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    def get(
        self,
        id: str,
    ) -> Checkout:
        """
        Get a checkout session by ID.

        **Scopes**: `checkouts:read` `checkouts:write`

        Args:
            id: The checkout session ID.

        Raises:
            ResourceNotFound: Checkout session not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    def update(
        self,
        id: str,
        **kwargs: typing.Unpack[CheckoutUpdate],
    ) -> Checkout:
        """
        Update a checkout session.

        **Scopes**: `checkouts:write`

        Args:
            id: The checkout session ID.

        Raises:
            Update403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkouts/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            403: Update403Error,
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    def client_get(
        self,
        client_secret: str,
    ) -> CheckoutPublic:
        """
        Get a checkout session by client secret.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/client/{client_secret}",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublic, method_errors)

    def client_update(
        self,
        client_secret: str,
        **kwargs: typing.Unpack[CheckoutUpdatePublic],
    ) -> CheckoutPublic:
        """
        Update a checkout session by client secret.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            ClientUpdate403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkouts/client/{client_secret}",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            403: ClientUpdate403Error,
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublic, method_errors)

    def client_confirm(
        self,
        client_secret: str,
        **kwargs: typing.Unpack[CheckoutConfirmStripe],
    ) -> CheckoutPublicConfirmed:
        """
        Confirm a checkout session by client secret.

        Orders and subscriptions will be processed.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            PaymentError: The payment failed.
            ClientConfirm403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkouts/client/{client_secret}/confirm",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            400: PaymentError,
            403: ClientConfirm403Error,
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublicConfirmed, method_errors)


class CheckoutsAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        product_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        status: CheckoutStatus | builtins.list[CheckoutStatus] | None = None,
        query: str | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[CheckoutSortProperty] | None = ["-created_at"],
    ) -> ListResourceCheckout:
        """
        List checkout sessions.

        **Scopes**: `checkouts:read` `checkouts:write`

        Args:
            organization_id: Filter by organization ID.
            product_id: Filter by product ID.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by customer external ID.
            status: Filter by checkout session status.
            query: Filter by customer email.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "product_id": product_id,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "status": status,
                "query": query,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceCheckout, method_errors)

    async def create(
        self,
        **kwargs: typing.Unpack[CheckoutCreate],
    ) -> Checkout:
        """
        Create a checkout session.

        **Scopes**: `checkouts:write`

        Args:

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkouts/",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    async def get(
        self,
        id: str,
    ) -> Checkout:
        """
        Get a checkout session by ID.

        **Scopes**: `checkouts:read` `checkouts:write`

        Args:
            id: The checkout session ID.

        Raises:
            ResourceNotFound: Checkout session not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    async def update(
        self,
        id: str,
        **kwargs: typing.Unpack[CheckoutUpdate],
    ) -> Checkout:
        """
        Update a checkout session.

        **Scopes**: `checkouts:write`

        Args:
            id: The checkout session ID.

        Raises:
            Update403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkouts/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            403: Update403Error,
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Checkout, method_errors)

    async def client_get(
        self,
        client_secret: str,
    ) -> CheckoutPublic:
        """
        Get a checkout session by client secret.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/checkouts/client/{client_secret}",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublic, method_errors)

    async def client_update(
        self,
        client_secret: str,
        **kwargs: typing.Unpack[CheckoutUpdatePublic],
    ) -> CheckoutPublic:
        """
        Update a checkout session by client secret.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            ClientUpdate403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/checkouts/client/{client_secret}",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            403: ClientUpdate403Error,
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublic, method_errors)

    async def client_confirm(
        self,
        client_secret: str,
        **kwargs: typing.Unpack[CheckoutConfirmStripe],
    ) -> CheckoutPublicConfirmed:
        """
        Confirm a checkout session by client secret.

        Orders and subscriptions will be processed.

        Args:
            client_secret: The checkout session client secret.

        Raises:
            PaymentError: The payment failed.
            ClientConfirm403Error: The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
            ResourceNotFound: Checkout session not found.
            ExpiredCheckoutError: The checkout session is expired.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/checkouts/client/{client_secret}/confirm",
            path_params={
                "client_secret": client_secret,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            400: PaymentError,
            403: ClientConfirm403Error,
            404: ResourceNotFound,
            410: ExpiredCheckoutError,
            422: HTTPValidationError,
        }
        return parse_response(response, CheckoutPublicConfirmed, method_errors)

File: sdk/typescript/src/services/checkouts.ts
import { ClientBase } from "../base";
import type {
  CheckoutConfirmStripe,
  CheckoutCreate,
  CheckoutUpdate,
  CheckoutUpdatePublic,
} from "../models/inputs";
import type {
  Checkout,
  CheckoutPublic,
  CheckoutPublicConfirmed,
  ListResourceCheckout,
} from "../models/outputs";
import type { CheckoutSortProperty, CheckoutStatus } from "../models/literals";
import {
  ClientConfirm403Error,
  ClientUpdate403Error,
  ExpiredCheckoutError,
  HTTPValidationError,
  PaymentError,
  ResourceNotFound,
  Update403Error,
} from "../errors";

export const listCheckouts = (client: ClientBase) => {
  /**
   * List checkout sessions.
   *
   * **Scopes**: `checkouts:read` `checkouts:write`
   *
   * @param query - Query parameters
   * @returns {ListResourceCheckout}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    organization_id?: string | string[] | null;
    product_id?: string | string[] | null;
    customer_id?: string | string[] | null;
    external_customer_id?: string | string[] | null;
    status?: CheckoutStatus | CheckoutStatus[] | null;
    query?: string | null;
    page?: number;
    limit?: number;
    sorting?: CheckoutSortProperty[] | null;
  }): Promise<ListResourceCheckout> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
      product_id: query?.product_id,
      customer_id: query?.customer_id,
      external_customer_id: query?.external_customer_id,
      status: query?.status,
      query: query?.query,
      page: query?.page || 1,
      limit: query?.limit || 10,
      sorting: query?.sorting || ["-created_at"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/checkouts/",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourceCheckout>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const createCheckouts = (client: ClientBase) => {
  /**
   * Create a checkout session.
   *
   * **Scopes**: `checkouts:write`
   *
   * @param body - Request body* @returns {Checkout}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body?: CheckoutCreate): Promise<Checkout> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest("POST", "/v1/checkouts/", pathParams, queryParams, body);
    const response = await client.sendRequest(request);
    return client.parseResponse<Checkout>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const getCheckouts = (client: ClientBase) => {
  /**
   * Get a checkout session by ID.
   *
   * **Scopes**: `checkouts:read` `checkouts:write`
   *
   * @param id - The checkout session ID.
   * @returns {Checkout}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Checkout session not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<Checkout> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/checkouts/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<Checkout>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const updateCheckouts = (client: ClientBase) => {
  /**
   * Update a checkout session.
   *
   * **Scopes**: `checkouts:write`
   *
   * @param id - The checkout session ID.
   * @param body - Request body* @returns {Checkout}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {Update403Error} The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
   * @throws {ResourceNotFound} Checkout session not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string, body?: CheckoutUpdate): Promise<Checkout> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/checkouts/{id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<Checkout>(response, "json", {
      403: Update403Error,
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const clientGetCheckouts = (client: ClientBase) => {
  /**
   * Get a checkout session by client secret.
   *
   * @param client_secret - The checkout session client secret.
   * @returns {CheckoutPublic}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Checkout session not found.
   * @throws {ExpiredCheckoutError} The checkout session is expired.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (client_secret: string): Promise<CheckoutPublic> => {
    const pathParams = {
      client_secret: client_secret,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/checkouts/client/{client_secret}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutPublic>(response, "json", {
      404: ResourceNotFound,
      410: ExpiredCheckoutError,
      422: HTTPValidationError,
    });
  };
};
export const clientUpdateCheckouts = (client: ClientBase) => {
  /**
   * Update a checkout session by client secret.
   *
   * @param client_secret - The checkout session client secret.
   * @param body - Request body* @returns {CheckoutPublic}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ClientUpdate403Error} The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
   * @throws {ResourceNotFound} Checkout session not found.
   * @throws {ExpiredCheckoutError} The checkout session is expired.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (client_secret: string, body?: CheckoutUpdatePublic): Promise<CheckoutPublic> => {
    const pathParams = {
      client_secret: client_secret,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/checkouts/client/{client_secret}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutPublic>(response, "json", {
      403: ClientUpdate403Error,
      404: ResourceNotFound,
      410: ExpiredCheckoutError,
      422: HTTPValidationError,
    });
  };
};
export const clientConfirmCheckouts = (client: ClientBase) => {
  /**
   * Confirm a checkout session by client secret.
   *
   * Orders and subscriptions will be processed.
   *
   * @param client_secret - The checkout session client secret.
   * @param body - Request body* @returns {CheckoutPublicConfirmed}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {PaymentError} The payment failed.
   * @throws {ClientConfirm403Error} The checkout is expired, the customer already has an active subscription, or the organization is not ready to accept payments.
   * @throws {ResourceNotFound} Checkout session not found.
   * @throws {ExpiredCheckoutError} The checkout session is expired.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    client_secret: string,
    body?: CheckoutConfirmStripe,
  ): Promise<CheckoutPublicConfirmed> => {
    const pathParams = {
      client_secret: client_secret,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "POST",
      "/v1/checkouts/client/{client_secret}/confirm",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<CheckoutPublicConfirmed>(response, "json", {
      400: PaymentError,
      403: ClientConfirm403Error,
      404: ResourceNotFound,
      410: ExpiredCheckoutError,
      422: HTTPValidationError,
    });
  };
};

export function createCheckoutsService(client: ClientBase) {
  return {
    list: listCheckouts(client),
    create: createCheckouts(client),
    get: getCheckouts(client),
    update: updateCheckouts(client),
    clientGet: clientGetCheckouts(client),
    clientUpdate: clientUpdateCheckouts(client),
    clientConfirm: clientConfirmCheckouts(client),
  };
}

export type Checkouts = ReturnType<typeof createCheckoutsService>;
